lib/src/string_scanner.dart

```diff
@@ -120,6 +139,45 @@ class StringScanner {
     _fail(name);
   }
 
+  /// Consumes a single Unicode code unit and returns it.
+  ///
+  /// This works like [readChar], except that it automatically handles UTF-16
+  /// surrogate pairs. Specifically, if the next character is a high surrogate,
+  /// it consumes it as well as the following character and returns the resolved
+  /// Unicode code point.
+  ///
+  /// If the scanner is looking at an unpaired surrogate, this returns the high
+  /// surrogate as-is.
+  int readCodePoint() {
+    final first = readChar();
+    if (!isHighSurrogate(first)) return first;
+
+    final next = peekChar();
+    if (next == null || !isLowSurrogate(next)) return first;
+
+    readChar();
+    return decodeSurrogatePair(first, next);
+  }
+
+  /// Returns the Unicode code point immediately after [position].
+  ///
+  /// This works like [peekChar], except that it automatically handles UTF-16
+  /// surrogate pairs. Specifically, if the next character is a high surrogate,
+  /// it consumes it as well as the following character and returns the resolved
```

Comment:
> ... is a high surrogate followed by a low surrogate, this method consumes both and returns the 
> corresponding Unicode code point.
>
> Otherwise this method consumes and returns just the next code unit, even if it's an unpaired surrogate.